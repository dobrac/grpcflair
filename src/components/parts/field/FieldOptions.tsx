import protobuf from "protobufjs";
import Options from "@/components/parts/helpers/Options";

export interface FieldOptionsProps {
  field: protobuf.Field | protobuf.OneOf;
  dark?: boolean;
}

export default function FieldOptions({ field, dark }: FieldOptionsProps) {
  return <Options reflectionObject={field} />;
}
