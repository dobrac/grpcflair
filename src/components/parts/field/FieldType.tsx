import protobuf, { MapField } from "protobufjs";
import Type from "@/components/parts/Type";
import { useState } from "react";
import { Button } from "react-bootstrap";
import EnumType from "@/components/parts/EnumType";
import FieldComment from "@/components/parts/field/FieldComment";

export interface FieldTypeProps {
  field: protobuf.Field;
  dark?: boolean;
  expanded?: boolean;
}

export default function FieldType({
  field,
  dark,
  expanded = false,
}: FieldTypeProps) {
  field.resolve();

  const [open, setOpen] = useState(expanded);

  let resolvedType = field.resolvedType;

  const expandable = !!resolvedType;

  return (
    <>
      {field.map && "map<" + (field as unknown as MapField).keyType + ", "}
      <span className={dark ? "text-white" : "text-secondary"}>
        {expandable ? (
          <Button
            size="sm"
            variant="link"
            className="px-0"
            onClick={() => setOpen((open) => !open)}
          >
            {field.type}
          </Button>
        ) : (
          field.type
        )}
      </span>
      {field.map && ">"}
      {field.repeated ? "[]" : ""}
      {/*{field.optional ? " (optional)" : ""}*/}
      <div>
        {!!resolvedType && resolvedType instanceof protobuf.Type && open && (
          <Type type={resolvedType} dark={dark} expanded={true} />
        )}
        {!!resolvedType && resolvedType instanceof protobuf.Enum && open && (
          <EnumType enumType={resolvedType} dark={dark} expanded={true} />
        )}
      </div>
    </>
  );
}
