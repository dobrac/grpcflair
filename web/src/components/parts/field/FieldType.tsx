import protobuf, { MapField } from "protobufjs";
import Type from "@/components/parts/Type";
import { useState } from "react";
import { Button } from "react-bootstrap";
import EnumType from "@/components/parts/EnumType";

export interface FieldTypeProps {
  field: protobuf.Field;
  dark?: boolean;
  expandable?: boolean;
  expanded?: boolean;
  className?: string;
}

export default function FieldType({
  field,
  dark,
  expandable = true,
  expanded = false,
  className,
}: FieldTypeProps) {
  field.resolve();

  const [open, setOpen] = useState(expanded);

  let resolvedType = field.resolvedType;

  const expandableResolved = expandable && !!resolvedType;

  return (
    <span className={className}>
      <span className={dark ? "text-white" : "text-secondary"}>
        {field.map && "map<" + (field as unknown as MapField).keyType + ", "}
        {expandableResolved ? (
          <Button
            size="sm"
            variant="link"
            className="px-0"
            onClick={() => setOpen((open) => !open)}
          >
            {field.type}
          </Button>
        ) : (
          field.type
        )}
        {field.map && ">"}
        {field.repeated ? "[]" : ""}
      </span>
      {/*{field.optional ? " (optional)" : ""}*/}
      <div>
        {!!resolvedType && resolvedType instanceof protobuf.Type && open && (
          <Type type={resolvedType} dark={dark} expanded={true} />
        )}
        {!!resolvedType && resolvedType instanceof protobuf.Enum && open && (
          <EnumType enumType={resolvedType} dark={dark} expanded={true} />
        )}
      </div>
    </span>
  );
}
