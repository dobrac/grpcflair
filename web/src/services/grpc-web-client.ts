import protobuf from "protobufjs";
import {
  ClientReadableStream,
  GrpcWebClientBase,
  Metadata,
  MethodDescriptor,
  MethodType,
} from "grpc-web";
import { GrpcWebFormat } from "@/types/grpc-web";

/**
 * Dummy class to be used in the MethodDescriptor.
 */
class DummyRPCType {
  constructor(...args: unknown[]) {}
}

/**
 * Options for a gRPC call.
 */
export interface GrpcWebOptions {
  metadata?: Metadata;
  format?: GrpcWebFormat;
}

/**
 * The response of a unary gRPC call.
 */
export interface UnaryResponse<MessageData extends object = object> {
  headers?: Metadata;
  trailers?: Metadata;
  data?: protobuf.Message<MessageData>[];
}

/**
 * makeGrpcCall makes a unary gRPC call to the given service and method.
 * @param hostname The hostname of the gRPC server.
 * @param service The protobuf service.
 * @param method The protobuf method.
 * @param typeEncode The protobuf type to encode the message.
 * @param typeDecode The protobuf type to decode the response.
 * @param message The message to send.
 * @param options The request options.
 */
export function makeGrpcCall<MessageData extends object = object>(
  hostname: string,
  service: protobuf.Service,
  method: protobuf.Method,
  typeEncode: protobuf.Type,
  typeDecode: protobuf.Type,
  message: protobuf.Message,
  options?: GrpcWebOptions,
): Promise<UnaryResponse<MessageData>> {
  const client = new GrpcWebClientBase({ format: options?.format });

  const methodPath = `${hostname}/${
    // Replace starting dot "."
    service.fullName.replace(".", "")
  }/${method.name}`;

  return new Promise((resolve, reject) => {
    let completeResponse: UnaryResponse<MessageData> = {};

    const unaryStream = client.rpcCall(
      methodPath,
      // Ignored, using protobufjs directly
      {},
      options?.metadata ?? {},
      new MethodDescriptor(
        method.name,
        MethodType.UNARY,
        // Ignored, using protobufjs directly
        DummyRPCType,
        // Ignored, using protobufjs directly
        DummyRPCType,
        () => {
          return typeEncode.encode(message).finish();
        },
        (bytes: Uint8Array) => {
          return typeDecode.decode(bytes);
        },
      ),
      (err, response: protobuf.Message<MessageData>) => {
        if (err) {
          reject(err);
        } else {
          completeResponse.data = [response];
        }
      },
    );

    unaryStream.on("metadata", (metadata) => {
      completeResponse.headers = metadata;
    });

    unaryStream.on("status", (status) => {
      completeResponse.trailers = status.metadata;
    });

    unaryStream.on("end", () => {
      resolve(completeResponse);
    });
  });
}

/**
 * makeGrpcServerStreamingCall makes a server streaming gRPC call to the given service and method.
 * @param hostname The hostname of the gRPC server.
 * @param service The protobuf service.
 * @param method The protobuf method.
 * @param typeEncode The protobuf type to encode the message.
 * @param typeDecode The protobuf type to decode the response.
 * @param message The message to send.
 * @param options The request options.
 */
export function makeGrpcServerStreamingCall<
  MessageData extends object = object,
>(
  hostname: string,
  service: protobuf.Service,
  method: protobuf.Method,
  typeEncode: protobuf.Type,
  typeDecode: protobuf.Type,
  message: protobuf.Message,
  options?: GrpcWebOptions,
): ClientReadableStream<protobuf.Message<MessageData>> {
  if (options?.format !== GrpcWebFormat.TEXT) {
    throw new Error(
      `Only format ${GrpcWebFormat.TEXT} is supported for grpc-web server streaming`,
    );
  }

  const client = new GrpcWebClientBase({ format: options?.format });

  const methodPath = `${hostname}/${
    // Replace starting dot "."
    service.fullName.replace(".", "")
  }/${method.name}`;

  return client.serverStreaming<{}, protobuf.Message<MessageData>>(
    methodPath,
    // Ignored, using protobufjs directly
    {},
    options?.metadata ?? {},
    new MethodDescriptor(
      method.name,
      MethodType.SERVER_STREAMING,
      // Ignored, using protobufjs directly
      DummyRPCType,
      // Ignored, using protobufjs directly
      DummyRPCType,
      () => {
        return typeEncode.encode(message).finish();
      },
      (bytes: Uint8Array) => {
        return typeDecode.decode(bytes);
      },
    ),
  );
}
