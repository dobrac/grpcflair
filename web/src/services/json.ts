/**
 * Check if a string is a valid JSON
 * @param str the string to check
 * @returns true if the string is a valid JSON, false otherwise
 */
export const isJSON = (str: string) => {
  try {
    JSON.parse(str);
  } catch (e) {
    return false;
  }
  return true;
};

/**
 * Transform the values of an object using a function
 * @param object the object to transform
 * @param transformValue the function to transform the values
 * @returns the transformed object
 */
export function transformObjectValues<T extends object>(
  object: T,
  transformValue: (value: unknown) => unknown,
): T {
  const result: any = {};
  for (const [key, value] of Object.entries(object)) {
    result[key] = transformValue(value);
  }
  return result;
}

/**
 * Clean empty (null or "") values from an object by setting them to undefined
 * @param obj the object to clean
 * @returns the cleaned object
 */
export function cleanEmptyValues<T extends object>(obj: T): T {
  return transformObjectValues(obj, (value) => {
    if (value === null) {
      return undefined;
    }
    if (value === "") {
      return undefined;
    }
    return value;
  });
}
