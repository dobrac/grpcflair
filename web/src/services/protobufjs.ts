import protobuf from "protobufjs";

/**
 * Returns the options from a reflection object.
 * @param reflectionObject the reflection object to get the options from
 * @param includeParent whether to include the parent's options
 * @returns the options from the reflection object
 */
export function getOptionsFromReflectionObject(
  reflectionObject: protobuf.ReflectionObject,
  includeParent = true,
): Record<string, unknown> {
  return {
    ...reflectionObject.options,
    ...(includeParent ? reflectionObject.parent?.options : {}),
  };
}

/**
 * Returns the services from a context.
 * @param context the context to get the services from
 * @returns the services from the context
 */
export function getServicesFromContext(
  context: protobuf.Root,
): protobuf.Service[] {
  const values: protobuf.Service[] = [];
  traverseStructure(context, isService, (type) => {
    values.push(type);
  });
  return values;
}

/**
 * Returns the types from a context.
 * @param context the context to get the types from
 * @returns the types from the context
 */
export function getTypesFromContext(context: protobuf.Root): protobuf.Type[] {
  const values: protobuf.Type[] = [];
  traverseStructure(context, isType, (type) => {
    values.push(type);
  });
  return values;
}

/**
 * Returns the enums from a context.
 * @param context the context to get the enums from
 * @returns the enums from the context
 */
export function getEnumsFromContext(context: protobuf.Root): protobuf.Enum[] {
  const values: protobuf.Enum[] = [];
  traverseStructure(context, isEnum, (type) => {
    values.push(type);
  });
  return values;
}

/**
 * Checks if a reflection object is a NamespaceBase type
 * @param current the reflection object to check
 * @returns true if the reflection object is a NamespaceBase, false otherwise
 */
function isNamespaceBase(
  current: protobuf.NamespaceBase | protobuf.ReflectionObject,
): current is protobuf.NamespaceBase {
  return (current as protobuf.NamespaceBase).nestedArray !== undefined;
}

type ProtobufJsType =
  | protobuf.Type
  | protobuf.Enum
  | protobuf.Service
  | protobuf.NamespaceBase
  | protobuf.ReflectionObject;

/**
 * Checks if a value is a protobuf.Type for TypeScript type guards
 * @param value the value to check
 */
function isType(value: ProtobufJsType): value is protobuf.Type {
  return value instanceof protobuf.Type;
}

/**
 * Checks if a value is a protobuf.Enum for TypeScript type guards
 * @param value the value to check
 */
function isEnum(value: ProtobufJsType): value is protobuf.Enum {
  return value instanceof protobuf.Enum;
}

/**
 * Checks if a value is a protobuf.Service for TypeScript type guards
 * @param value the value to check
 */
function isService(value: ProtobufJsType): value is protobuf.Service {
  return value instanceof protobuf.Service;
}

/**
 * Traverses a structure and calls a function on each type that matches the typeCheckFn
 *
 * Source: https://github.com/protobufjs/protobuf.js/blob/master/examples/traverse-types.js
 *
 * @param current the current type to traverse
 * @param typeCheckFn the function to check if the type matches
 * @param fn the function to call on the type
 */
function traverseStructure<DesiredType extends ProtobufJsType>(
  current: ProtobufJsType,
  typeCheckFn: (value: ProtobufJsType) => value is DesiredType,
  fn: (type: DesiredType) => void,
) {
  if (typeCheckFn(current)) {
    // and/or protobuf.Enum, protobuf.Service etc.
    fn(current);
  }
  if (isNamespaceBase(current)) {
    current.nestedArray.forEach(function (nested) {
      traverseStructure(nested, typeCheckFn, fn);
    });
  }
}

export enum RequestType {
  UNARY,
  CLIENT_STREAMING,
  SERVER_STREAMING,
  BIDIRECTIONAL_STREAMING,
}

/**
 * Get the request type of method
 * @param method the method to get the request type from
 * @returns the request type of the method
 */
export function getRequestType(method: protobuf.Method): RequestType {
  if (method.requestStream && method.responseStream) {
    return RequestType.BIDIRECTIONAL_STREAMING;
  } else if (method.requestStream) {
    return RequestType.CLIENT_STREAMING;
  } else if (method.responseStream) {
    return RequestType.SERVER_STREAMING;
  } else {
    return RequestType.UNARY;
  }
}

/**
 * Get the color of a method based on its type
 * @param method the method to get the color from
 * @returns the color of the method
 */
export function getColorFromMethodType(method: protobuf.Method): string {
  switch (getRequestType(method)) {
    case RequestType.BIDIRECTIONAL_STREAMING:
    case RequestType.CLIENT_STREAMING:
      return "danger";
    case RequestType.SERVER_STREAMING:
      return "dark";
    case RequestType.UNARY:
      return "success";
  }
}

/**
 * Get the display name of a request type
 * @param type the request type to get the display name from
 * @returns the display name of the request type
 */
export function getRequestTypeDisplayName(type: RequestType) {
  switch (type) {
    case RequestType.BIDIRECTIONAL_STREAMING:
      return "Bi-directional streaming";
    case RequestType.CLIENT_STREAMING:
      return "Client streaming";
    case RequestType.SERVER_STREAMING:
      return "Server streaming";
    case RequestType.UNARY:
      return "Unary";
  }
}

/**
 * Get the display name of a method type
 * @param method the method to get the display name from
 * @returns the display name of the method type
 */
export function getMethodTypeDisplayName(method: protobuf.Method) {
  return getRequestTypeDisplayName(getRequestType(method));
}

/**
 * Transform the values of an object using a function
 * @param type the type object to transform
 * @param transformValue the function to transform the values
 * @returns the transformed object
 */
export function transformTypeValues(
  type: protobuf.Type | null,
  transformValue: (value: protobuf.Field) => unknown,
) {
  const result: any = {};
  if (type == null) {
    return result;
  }

  for (const field of type.fieldsArray) {
    field.resolve();
    const fieldName = field.name;

    result[fieldName] = transformValue(field);
  }
  return result;
}

export const exportedForTesting = {
  isNamespaceBase,
  isType,
  isEnum,
  isService,
  traverseStructure,
};
