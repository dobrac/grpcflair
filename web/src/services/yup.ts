import protobuf from "protobufjs";
import * as yup from "yup";

/**
 * Returns the yup type for a field.
 * @param field the field to get the yup type for
 * @returns the yup type for the field
 */
export function getFieldYupType(field: protobuf.Field): yup.Schema {
  if (field.map) {
    return yup
      .object()
      .typeError(`Field "${field.name}" must be an object`)
      .transform((val, orig) => (orig == "" ? null : val))
      .nullable()
      .json();
  }

  if (field.repeated) {
    return yup
      .array()
      .typeError(`Field "${field.name}" must be an array object`)
      .transform((val, orig) => (orig == "" ? null : val))
      .nullable()
      .json();
  }

  if (field.resolvedType == null) {
    if (typeof field.typeDefault === "boolean") {
      return yup
        .boolean()
        .typeError(`Field "${field.name}" must be a boolean`)
        .transform((val, orig) => (orig == "" ? null : val))
        .nullable();
    }

    if (typeof field.typeDefault === "number") {
      return yup
        .number()
        .typeError(`Field "${field.name}" must be a number`)
        .transform((val, orig) => (orig == "" ? null : val))
        .nullable();
    }

    if (field.type === "bytes") {
      return yup.mixed<Uint8Array>().nullable();
    }

    return yup.string().nullable();
  }

  if (field.resolvedType instanceof protobuf.Enum) {
    return yup
      .number()
      .transform((val, orig) => (orig == "" ? null : val))
      .nullable();
  }

  return yup
    .object()
    .transform((val, orig) => (orig == "" ? null : val))
    .typeError(`Field "${field.name}" must be an object`)
    .nullable()
    .json();
}
