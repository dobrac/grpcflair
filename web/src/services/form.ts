import protobuf, { MapField } from "protobufjs";
import { transformTypeValues } from "@/services/protobufjs";

/**
 * MAX_DEPTH is the maximum depth to traverse the protobuf message to generate
 */
const MAX_DEPTH = 1;

/**
 * getFieldPlaceholderValue returns a placeholder value for a field.
 * If the field is a message, it will return a placeholder object.
 * If the field is an enum, it will return the first enum value.
 * If the field is a scalar, it will return the default value or the type.
 * @param depth the depth of the field
 * @param field the protobuf field
 * @returns a placeholder value
 */
function getFieldPlaceholderValue(
  depth: number,
  field: protobuf.Field,
): unknown {
  if (field.resolvedType == null) {
    const value = field.typeDefault ?? field.type;

    // For empty strings, return type name
    if (value === "") {
      return field.type;
    }

    return value;
  }

  if (field.resolvedType instanceof protobuf.Enum) {
    return Object.values(field.resolvedType.values)[0];
  }

  if (depth >= MAX_DEPTH) {
    return {};
  }

  return transformTypeValues(
    field.resolvedType,
    placeholderTransformationWithDepth.bind(null, depth + 1),
  );
}

/**
 * placeholderTransformationWithDepth returns a placeholder transformation for a field.
 * If the field is repeated, it will return an array with a placeholder value.
 * If the field is a map, it will return an object with a placeholder key and value.
 * If the field is a scalar, it will return a placeholder value.
 * @param depth the depth of the field
 * @param field the protobuf field
 * @returns a placeholder transformation
 */
function placeholderTransformationWithDepth(
  depth: number,
  field: protobuf.Field,
) {
  if (field.repeated) {
    return [getFieldPlaceholderValue(depth, field)];
  }

  if (field.map) {
    const fieldMap = field as unknown as MapField;
    const fieldMapKeyValue = fieldMap.keyType;
    const value = getFieldPlaceholderValue(depth, field);
    return {
      [fieldMapKeyValue]: value,
    };
  }

  return getFieldPlaceholderValue(depth, field);
}

/**
 * placeholderTransformation returns a placeholder transformation for a field.
 * If the field is repeated, it will return an array with a placeholder value.
 * If the field is a map, it will return an object with a placeholder key and value.
 * If the field is a scalar, it will return a placeholder value.
 * @param field the protobuf field
 * @returns a placeholder transformation
 */
export function placeholderTransformation(field: protobuf.Field) {
  return placeholderTransformationWithDepth(0, field);
}

/**
 * formTransformation returns a JSON string of the placeholder transformation for a field.
 * If the field is repeated or a map, it will return an empty string.
 * If the field is an enum, it will return an empty string.
 * If the field is a message, it will return a JSON string of the placeholder transformation.
 * @param field the protobuf field
 * @returns a JSON string of the placeholder transformation
 */
export function formTransformation(field: protobuf.Field): string {
  if (field.repeated || field.map) {
    return "";
  }

  if (field.resolvedType == null) {
    return "";
  }

  if (field.resolvedType instanceof protobuf.Enum) {
    return "";
  }

  return JSON.stringify(
    transformTypeValues(field.resolvedType, placeholderTransformation),
    null,
    2,
  );
}

export const exportedForTesting = {
  getFieldPlaceholderValue,
  placeholderTransformationWithDepth,
};
