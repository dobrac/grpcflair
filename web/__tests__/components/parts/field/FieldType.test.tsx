import { act, render, screen } from "@testing-library/react";
import { type } from "../../../../tests/protobufjs-source";
import { getOptionsFromReflectionObject } from "@/services/protobufjs";
import FieldOptions from "@/components/parts/field/FieldOptions";
import FieldType from "@/components/parts/field/FieldType";
import protobuf from "protobufjs";

describe("FieldType", () => {
  type.fieldsArray.forEach((field) => {
    field.resolve();

    it(`renders ${field.name} - expanded false`, async () => {
      const renderedFieldType = await act(async () => {
        return render(
          <FieldType field={field} expanded={false} expandable={false} />,
        );
      });

      const element = renderedFieldType.container.firstChild;
      expect(element).toBeInTheDocument();

      expect(element).toMatchSnapshot();
    });

    it(`renders ${field.name} - expanded true`, async () => {
      const renderedFieldType = await act(async () => {
        return render(
          <FieldType field={field} expanded={true} expandable={true} />,
        );
      });

      const element = renderedFieldType.container.firstChild;
      expect(element).toBeInTheDocument();

      const typeDetail = renderedFieldType.queryByTestId("type-detail");
      if (field.resolvedType instanceof protobuf.Type) {
        expect(typeDetail).toBeInTheDocument();
      } else {
        expect(typeDetail).not.toBeInTheDocument();
      }

      const enumDetail = renderedFieldType.queryByTestId("enum");
      if (field.resolvedType instanceof protobuf.Enum) {
        expect(enumDetail).toBeInTheDocument();
      } else {
        expect(enumDetail).not.toBeInTheDocument();
      }
    });
  });
});
