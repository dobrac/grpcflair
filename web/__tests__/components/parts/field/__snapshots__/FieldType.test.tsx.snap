// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldType renders any - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    google.protobuf.Any
  </span>
  <div />
</span>
`;

exports[`FieldType renders bool - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    bool
  </span>
  <div />
</span>
`;

exports[`FieldType renders bytes - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    bytes
  </span>
  <div />
</span>
`;

exports[`FieldType renders enum - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    Corpus
  </span>
  <div />
</span>
`;

exports[`FieldType renders enumAlias - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    EnumAllowingAlias
  </span>
  <div />
</span>
`;

exports[`FieldType renders map - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    map&lt;string, 
    google.protobuf.Any
    &gt;
  </span>
  <div />
</span>
`;

exports[`FieldType renders number - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    int32
  </span>
  <div />
</span>
`;

exports[`FieldType renders oneofNumber_1 - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    int32
  </span>
  <div />
</span>
`;

exports[`FieldType renders oneofNumber_2 - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    int32
  </span>
  <div />
</span>
`;

exports[`FieldType renders oneofString - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    string
  </span>
  <div />
</span>
`;

exports[`FieldType renders repeatedEnum - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    Corpus
    []
  </span>
  <div />
</span>
`;

exports[`FieldType renders repeatedNumber - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    int32
    []
  </span>
  <div />
</span>
`;

exports[`FieldType renders repeatedString - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    string
    []
  </span>
  <div />
</span>
`;

exports[`FieldType renders string - expanded false 1`] = `
<span>
  <span
    class="text-secondary"
  >
    string
  </span>
  <div />
</span>
`;
